rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1652 @@
+    // }
+
+    // async fn call_unobserve_asset_update(
+    //     device_name: &str,
+    //     endpoint_name: &str,
+    //     asset_name: &str,
+    //     timeout: Duration,
+    //     adr_client: Client<SessionManagedClient>,
+    // ) -> Result<(), Error> {
+    //     adr_client
+    //         .unobserve_asset_update_notifications(
+    //             device_name.to_string(),
+    //             endpoint_name.to_string(),
+    //             asset_name.to_string(),
+    //             timeout,
+    //         )
+    //         .await
+    // }
+
+    // async fn call_get_device(
+    //     device_name: &str,
+    //     endpoint_name: &str,
+    //     timeout: Duration,
+    //     adr_client: Client<SessionManagedClient>,
+    // ) -> Result<Device, Error> {
+    //     adr_client
+    //         .get_device(device_name.to_string(), endpoint_name.to_string(), timeout)
+    //         .await
+    // }
+
+    // async fn call_update_device_plus_endpoint(
+    //     device_name: &str,
+    //     endpoint_name: &str,
+    //     timeout: Duration,
+    //     adr_client: Client<SessionManagedClient>,
+    // ) -> Result<Device, Error> {
+    //     adr_client
+    //         .update_device_plus_endpoint_status(
+    //             device_name.to_string(),
+    //             endpoint_name.to_string(),
+    //             DeviceStatus::default(),
+    //             timeout,
+    //         )
+    //         .await
+    // }
+
+    // async fn call_observe_device_update(
+    //     device_name: &str,
+    //     endpoint_name: &str,
+    //     timeout: Duration,
+    //     adr_client: Client<SessionManagedClient>,
+    // ) -> Result<DeviceUpdateObservation, Error> {
+    //     adr_client
+    //         .observe_device_update_notifications(
+    //             device_name.to_string(),
+    //             endpoint_name.to_string(),
+    //             timeout,
+    //         )
+    //         .await
+    // }
+
+    // async fn call_unobserve_device_update(
+    //     device_name: &str,
+    //     endpoint_name: &str,
+    //     timeout: Duration,
+    //     adr_client: Client<SessionManagedClient>,
+    // ) -> Result<(), Error> {
+    //     adr_client
+    //         .unobserve_device_update_notifications(
+    //             device_name.to_string(),
+    //             endpoint_name.to_string(),
+    //             timeout,
+    //         )
+    //         .await
+    // }
+
+    #[tokio::test]
+    async fn test_get_asset_empty_asset_name() {
+        let adr_client = create_adr_client();
+        let result = adr_client
+            .get_asset(
+                "test-device".to_string(),
+                "test-endpoint".to_string(),
+                "".to_string(),
+                Duration::from_secs(10),
+            )
+            .await;
+
+        assert!(matches!(
+            result.unwrap_err().0,
+            ErrorKind::AIOProtocolError(ref e) if matches!(e.kind, AIOProtocolErrorKind::ClientError)
```

Comment:
ClientError should only be for network failures, is this the right behavior to validate?

Reply:
Not network failures, it's due to literal client infrastructure errors (i.e. session dropped so no event loop to send MQTT requests to)

Reply:
changed into a separate error type as the request gets simply dropped by the service